modules/ca/src/tools/caget.c

```diff
@@ -549,14 +626,36 @@ int main (int argc, char *argv[])
     for (n = 0; optind < argc; n++, optind++)
         pvs[n].name = argv[optind] ;       /* Copy PV names from command line */
 
-    result = connect_pvs(pvs, nPvs);
+    // Mutex to avoid commingling of text output by several threads
+    epicsMutexId printMutex = epicsMutexCreate();
+    // Mutex to access result counter
+    epicsMutexId resultMutex = epicsMutexCreate();
+    for (int iii=0; iii<nPvs; ++iii) {
+        t_thread_data* thread_data;
+        thread_data = malloc(sizeof(t_thread_data));
+        thread_data->pv_data = pvs[iii];
+        thread_data->request = request;
+        thread_data->format = format;
+        thread_data->dbrType = type;
+        thread_data->reqElems = count;
+        thread_data->printMutex = printMutex;
+        thread_data->resultMutex = resultMutex;
+        thread_builder(thread_data, iii);
+    }
 
-                                /* Read and print data */
-    if (!result)
-        result = caget(pvs, nPvs, request, format, type, count);
+    epicsThreadId thread_id;
+    char th_name[9];
+
+    // Wait for all threads to finish their business before ending the main thread
+    for (int iii=0; iii<nPvs; ++iii) {
+        thread_name(iii, th_name);
+        thread_id = epicsThreadGetId(th_name);
+        epicsThreadMustJoin(thread_id);
+    }
 
-                                /* Shut down Channel Access */
     ca_context_destroy();
+    epicsMutexDestroy(printMutex);
+    epicsMutexDestroy(resultMutex);
 
-    return result;
+    return pvs_with_trouble;
```

Comment:
I'm not convinced that this return status is a good idea. The C++ standards says this about `main()`:
> Execution of the return (or the implicit return upon reaching the end of main) is equivalent to first leaving the function normally (which destroys the objects with automatic storage duration) and then calling [std::exit](https://en.cppreference.com/w/cpp/utility/program/exit) with the same argument as the argument of the [return](https://en.cppreference.com/w/cpp/language/return) ([std::exit](https://en.cppreference.com/w/cpp/utility/program/exit) then destroys static objects and terminates the program).

and this abount the exit status value:
> If exit_code is 0 or [EXIT_SUCCESS](https://en.cppreference.com/w/cpp/utility/program/EXIT_status), an implementation-defined status indicating successful termination is returned. If exit_code is [EXIT_FAILURE](https://en.cppreference.com/w/cpp/utility/program/EXIT_status), an implementation-defined status indicating unsuccessful termination is returned. In other cases implementation-defined status value is returned.